library/common.h

```diff
@@ -68,6 +74,19 @@ extern void (*mbedtls_test_hook_test_fail)( const char * test, int line, const c
  */
 #define MBEDTLS_ALLOW_PRIVATE_ACCESS
 
+/** Detect architectures where unaligned memory accesses are safe and performant.
+ *
+ * This list is incomplete.
+ */
+#if (defined(__i386__) || defined(__amd64__) || defined( __x86_64__) \
+    || defined(__ARM_FEATURE_UNALIGNED) \
+    || defined(__aarch64__) \
+    || defined(__ARM_ARCH_8__) || defined(__ARM_ARCH_8A__) || defined(__ARM_ARCH_8M__) \
+    || defined(__ARM_ARCH_7A__)) \
+    && (!(defined(__has_feature) && __has_feature(undefined_behavior_sanitizer)))
+#define MBEDTLS_ALLOW_UNALIGNED_ACCESS
+#endif
```

Comment:
This is getting quite complicated now. I wonder about allowing the user to override this simply, by allowing them to define it on the compiler command line, and, if thus defined, we don't enable it.

The simple way to do this is to go from it being defined or not to being defined to 0 or 1:

```
/* Allow user override by defining to 0 or 1 on cc command line */
#if !defined(MBEDTLS_ALLOW_UNALIGNED_ACCESS)
#    if (defined(__i386__) || defined(__amd64__) || defined( __x86_64__) \
        || defined(__ARM_FEATURE_UNALIGNED) \
        || defined(__aarch64__) \
        || defined(__ARM_ARCH_8__) || defined(__ARM_ARCH_8A__) || defined(__ARM_ARCH_8M__) \
        || defined(__ARM_ARCH_7A__)) \
        && (!(defined(__has_feature) && __has_feature(undefined_behavior_sanitizer)))
#        define MBEDTLS_ALLOW_UNALIGNED_ACCESS 1
#    else
#        define MBEDTLS_ALLOW_UNALIGNED_ACCESS 0
#    endif
#endif
```

and then

```
#if defined(MBEDTLS_ALLOW_UNALIGNED_ACCESS)
```

below becomes

```
#if MBEDTLS_ALLOW_UNALIGNED_ACCESS

```

Reply:
Agreed, it was also upsetting the CI. I've tidied it up (slightly differently).

Reply:
Sorry, I wasn't very clear. It wasn't just about the sanitiser part of the expression, it was about giving users a simple way to either forcibly enable or forcibly disable this without having to add an additional macro in the config file. That is, by changing from having "defined or not defined" of `MBEDTLS_ALLOW_UNALIGNED_ACCESS` to enable/disable, have `MBEDTLS_ALLOW_UNALIGNED_ACCESS` be given a value of `1` or `0`, but only do this within our code if it's not already defined.

This allows a user to override this simply on the compiler command line, with either `-D MBEDTLS_ALLOW_UNALIGNED_ACCESS=0` or `-D MBEDTLS_ALLOW_UNALIGNED_ACCESS=1`.

@gilles-peskine-arm thoughts?

Reply:
My thoughts: the fewer compilation options the better. I would prefer not to have a bug that's only when input has certain alignment, only on X platform, only with X compiler when optimizing with -OX, _and only with X compilation option_. That list of things that can go wrong is too long already.

I'd prefer to either write standard C with no platform assumption, or assembly that gets included whenever it's available for the platform and `MBEDTLS_HAVE_ASM` is enabled (no option to pick and choose amongst the bits of assembly).

Reply:
There is no "standard C" way to do this as far as I know, even under C11. So the options are:

1. use the `aligned` attribute
2. use the `packed` attribute as per e.g. https://api.riot-os.org/unaligned_8h_source.html
3. use assembly
4. drop the PR

Under options 1 and 2, it's safe to simplify by dropping the architecture detection (but this would become a small perf regression on older architectures, so I haven't done this).

I think option 2 has no benefits over option 1 and is messier so I don't favour this.

Option 4 seems like the wrong choice to me - unaligned access is a reasonable feature to have in the codebase.

I think an attribute that the compiler understands is safer and more portable than assembly. It also lets us test it under ASan, which I don't think could be done with assembly.

So option 1 remains my preference, but I'm open to modifying by
- removing architecture detection (although I think it's better to keep it to avoid a perf regression)
- limiting to compilers that we have actually tested and proven to work with ASan or similar (so far, that's just clang)

Reply:
I agree this is a worthwhile speed improvement with low code-size cost. Of those options I prefer 1, for pretty much the same reasons. However, I do still think the option to override on the command line (i.e. using `0`/`1` rather than undefined/defined) would be useful.

I don't think those two modifications are necessary

Reply:
> may risk a performance regression

The alternative is a similar risk of correctness regression, which makes it a clear choice to me.

Reply:
I have a slight preference for the `memcpy()` version

Reply:
OK, absent further input I will rework as `memcpy()`, with no compiler detection.